File: Java/src/main/java/org/onlineoptimisation/problem/benchmark/BuecheRastriginFunction.java
package org.onlineoptimisation.problem.benchmark;

import org.armadillojava.Arma;
import org.armadillojava.Col;
import org.armadillojava.Datum;
import org.armadillojava.Op;

public class BuecheRastriginFunction extends BenchmarkProblem {

  public BuecheRastriginFunction(int numberOfDimensions) {
    super(numberOfDimensions);
  }

  @Override
  public double getObjectiveValueImplementation(Col parameter) {
    Col z = oscillate(parameter);
    
    Col scaling = scaling(Math.pow(10, 0.5));
    for(int n = 0; n < scaling.n_elem; n += 2) {
      if(z.at(n) > 0) {
        scaling.at(n, Op.TIMES, 10);
      }
    }
    z = z.elemTimes(scaling);
    
    return 10 * (_numberOfDimensions - Arma.accu(Arma.cos(z.times(2 * Datum.pi)))) + Arma.accu(Arma.square(z));
  }
  
  @Override
  public double getSoftConstraintValueImplementation(Col parameter) {
    return 100 * penality(parameter);
  }

}

File: src/optimisationProblem/benchmark/buecheRastriginFunction.cpp
#include <hop_bits/optimisationProblem/benchmark/buecheRastriginFunction.hpp>

#include <cmath>

namespace hop {
  BuecheRastriginFunction::BuecheRastriginFunction(const unsigned int &numberOfDimensions) : BenchmarkProblem(numberOfDimensions), _scaling(getScaling(sqrt(10.0))) {

  }

  void BuecheRastriginFunction::setTranslation(const arma::Col<double> &translation) {
    _translation = translation;
    for (std::size_t n = 0; n < _translation.n_elem; n += 2) {
      _translation.at(n) = std::abs(_translation.at(n));
    }
  }

  double BuecheRastriginFunction::getObjectiveValueImplementation(const arma::Col<double> &parameter) const {
    arma::Col<double> z = _scaling % getOscillationTransformation(parameter - _translation);
    for (std::size_t n = 0; n < z.n_elem; n += 2) {
      if (z.at(n) > 0.0) {
        z.at(n) *= 10.0;
      }
    }

    return 10.0 * (static_cast<double>(_numberOfDimensions) - arma::accu(arma::cos(2.0 * arma::datum::pi * z))) + std::pow(arma::norm(z), 2) + 100.0 * getPenality(parameter);
  }
}
